### PR TITLE
ESP8266 & ESP32 without SPIFFS

### PR DESCRIPTION
**Pull request by Onno-Dirkzwager**
Hi Nredor,
First of thankyou for your previous work in making this esp8266 compatible and available as a library!

I started out with the mission to make this library work with hardware serial on the ESP32 platform. In the end I went all in and might have taken it to far for a pull request...... If you do accept I'll help you maintain it!

- Works with ESP32's
- This is more stable uploading larger files. (previously 478 kB updates would fail more often than not)
- Works with byte [Streams or file buffers](https://github.com/Onno-Dirkzwager/ESPNexUpload/blob/master/README.md) instead of a SPIFFS file
- Allows larger files then SPIFFS can hold.
- 115200 runs stable.
- faster as it skips spiffs and runs at full speed.
- 4 new [examples](https://github.com/Onno-Dirkzwager/ESPNexUpload/tree/master/examples)!

**UploadServer new**
- 318544 bytes (30%) of program storage space
- 30620 bytes (37%) of dynamic memory
- Upload time:  50.12s | Testfile: 478 kB (490.355 bytes) | Works 10 out of 10 times!

**UploadServer old**
- 317964 bytes (30%) of program storage space
- 31224 bytes (38%) of dynamic memory
- Upload time:  1m 25.84s | Testfile: 478 kB (490.355 bytes)

**Future**
- Add a "get info" function to get the model name, flash size etc.
- I would like to make it compatible with the Arduino Mkr Wifi series.
- Maby support esp8266 hardware serial (send serial monitor over softserial...)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,14 +1,55 @@
...
 ## Usage
-See example. Library expects that given file path is found at ESP8266 flash memory(SPIFFS). Included example provides webserver with file upload to SPIFFS. Sometimes upload doesnt work -> try to power cycle display.
+This library expects a byte (buffer of Stream), filesize and serial baudrate as parameters. The examples provided demonstrate how to use these with Wifi, SPIFFS, SD cards and a local webserver. 
+
+#### Begin
+```C++
+  // initialize ESPNexUpload
+  ESPNexUpload nextion(115200);
+  
+  // prepair upload: setup serial connection, send update command and send the expected update size
```
Comment (Nredor) [inline README.md]:
```suggestion
  // prepare upload: setup serial connection, send update command and send the expected update size
```

File: README.md [hunk 2]
```diff
@@ -1,14 +1,55 @@
...
+  // returns a True or False.
+  nextion.prepairUpload(fileSize);
```
Comment (Nredor) [inline README.md]: prepare* 
```suggestion
  nextion.prepareUpload(fileSize);
```

File: examples/HttpClient/HttpClient.ino [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+      // initialize ESPNexUpload
+      ESPNexUpload nextion(115200);
+
+      // set callback: What to do / show during upload..... Optional!
+      nextion.setUpdateProgressCallback([](){
+        Serial.print(".");
+      });
+      
+      // prepair upload: setup serial connection, send update command and send the expected update size
```
Comment (Nredor) [inline examples/HttpClient/HttpClient.ino]:
```suggestion
      // prepare upload: setup serial connection, send update command and send the expected update size
```

File: examples/HttpClient/HttpClient.ino [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+      result = nextion.prepairUpload(contentLength);
```
Comment (Nredor) [inline examples/HttpClient/HttpClient.ino]:
```suggestion
      result = nextion.prepareUpload(contentLength);
```

File: examples/SDRead/SDRead.ino [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+      
+      // initialize ESPNexUpload
+      ESPNexUpload nextion(115200);
+
+      // set callback: What to do / show during upload.... Optional!
+      nextion.setUpdateProgressCallback([](){
+        Serial.print(".");
+      });
+      
+      // prepair upload: setup serial connection, send update command and send the expected update size
```
Comment (Nredor) [inline examples/SDRead/SDRead.ino]:
```suggestion
      // prepare upload: setup serial connection, send update command and send the expected update size
```

File: examples/SDRead/SDRead.ino [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+      result = nextion.prepairUpload(myFile.size());
```
Comment (Nredor) [inline examples/SDRead/SDRead.ino]:
```suggestion
      result = nextion.prepareUpload(myFile.size());
```

File: examples/SPIFFSRead/SPIFFSRead.ino [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+      
+      // initialize ESPNexUpload
+      ESPNexUpload nextion(115200);
+
+      // set callback: What to do / show during upload.... Optional!
+      nextion.setUpdateProgressCallback([](){
+        Serial.print(".");
+      });
+      
+      // prepair upload: setup serial connection, send update command and send the expected update size
```
Comment (Nredor) [inline examples/SPIFFSRead/SPIFFSRead.ino]:
```suggestion
      // prepare upload: setup serial connection, send update command and send the expected update size
```

File: examples/SPIFFSRead/SPIFFSRead.ino [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+      result = nextion.prepairUpload(myFile.size());
```
Comment (Nredor) [inline examples/SPIFFSRead/SPIFFSRead.ino]:
```suggestion
      result = nextion.prepareUpload(myFile.size());
```

File: examples/UploadServer/UploadServer.ino [hunk 9]
```diff
@@ -1,117 +1,149 @@
...
   }
-        
+
+
   if(upload.status == UPLOAD_FILE_START){
-    // String filename = upload.filename;
+
+    Serial.println(F("\nFile received. Update Nextion..."));
+
+    // Prepair the Nextion display by seting up serial and telling it the file size to expect
```
Comment (Nredor) [inline examples/UploadServer/UploadServer.ino]:
```suggestion
    // Prepare the Nextion display by seting up serial and telling it the file size to expect
```

File: examples/UploadServer/UploadServer.ino [hunk 10]
```diff
@@ -1,117 +1,149 @@
...
+    result = nextion.prepairUpload(fileSize);
```
Comment (Nredor) [inline examples/UploadServer/UploadServer.ino]:
```suggestion
    result = nextion.prepareUpload(fileSize);
```

File: examples/WifiClient/WifiClient.ino [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+      // initialize ESPNexUpload
+      ESPNexUpload nextion(115200);
+
+      // set callback: What to do / show during upload.... Optional!
+      nextion.setUpdateProgressCallback([](){
+        Serial.print(".");
+      });
+      
+      // prepair upload: setup serial connection, send update command and send the expected update size
```
Comment (Nredor) [inline examples/WifiClient/WifiClient.ino]:
```suggestion
      // prepare upload: setup serial connection, send update command and send the expected update size
```

File: examples/WifiClient/WifiClient.ino [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+      result = nextion.prepairUpload(contentLength);
```
Comment (Nredor) [inline examples/WifiClient/WifiClient.ino]:
```suggestion
      result = nextion.prepareUpload(contentLength);
```

File: keywords.txt [hunk 13]
```diff
@@ -1,2 +1,3 @@
 ESPNexUpload	KEYWORD1
-upload	KEYWORD2
\ No newline at end of file
+upload	KEYWORD2
+prepairUpload	KEYWORD2
```
Comment (Nredor) [inline keywords.txt]:
```suggestion
prepareUpload	KEYWORD2
```

File: src/ESPNexUpload.h [hunk 14]
```diff
@@ -30,44 +35,99 @@
...
+     * Connect to Nextion over serial
+     *
+     * @return true or false.
+     */
+    bool connect();
     
-    /*
-     * start download.
+    /**
+     * prepair upload. Set file size & Connect to Nextion over serial
```
Comment (Nredor) [inline src/ESPNexUpload.h]:
```suggestion
     * prepare upload. Set file size & Connect to Nextion over serial
```

File: src/ESPNexUpload.h [hunk 15]
```diff
@@ -30,44 +35,99 @@
...
      *
      * @return true if success, false for failure.
      */
-    bool upload();
-    bool upload(String &statusMessage);
+	bool prepairUpload(uint32_t file_size);
```
Comment (Nredor) [inline src/ESPNexUpload.h]:
```suggestion
	bool prepareUpload(uint32_t file_size);
```

File: src/ESPNexUpload.cpp [hunk 16]
```diff
@@ -28,230 +28,338 @@
...
-    return true;
+	
+	return true;
 }
 
-uint16_t ESPNexUpload::_getBaudrate(void)
-{
+
+
+bool ESPNexUpload::prepairUpload(uint32_t file_size){
```
Comment (Nredor) [inline src/ESPNexUpload.cpp]:
```suggestion
bool ESPNexUpload::prepareUpload(uint32_t file_size){
```

**Comment by Onno-Dirkzwager** [comment]
Hi Nredor,
thanks voor fixing my typo's. I don't know if it is me but after aproving your suggested changes I end up with Changes requested: 1 review requesting changes.

Any tips on getting out of this mess?

**Pull request merged**